File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/place/ResidencyLetterDtos.kt
package app.pantopus.android.data.api.models.place

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * DTOs for `/api/homes/:id/residency-letters*` and the public
 * third-party check `/api/public/residency-letters/:code`. Route:
 * `backend/routes/residencyLetters.js` (serializeLetter at :185).
 * Mirrors `frontend/packages/api/src/endpoints/residencyLetters.ts`
 * and the iOS `ResidencyLetterDTOs.swift`.
 *
 * A T4 (verified-occupancy) resident issues a letter; the backend
 * freezes the printed facts + the exact PDF, and prints an unguessable
 * verification code on the letter. Letters are PERSONAL documents —
 * the API only ever returns the caller's own letters for a home.
 */

enum class ResidencyLetterStatus {
    @Json(name = "issued")
    ISSUED,

    @Json(name = "revoked")
    REVOKED,
    UNKNOWN,
}

@JsonClass(generateAdapter = true)
data class ResidencyLetterAddress(
    val line1: String,
    val city: String? = null,
    val state: String? = null,
    val zipcode: String? = null,
)

@JsonClass(generateAdapter = true)
data class ResidencyLetter(
    val id: String,
    @Json(name = "home_id") val homeId: String,
    val status: ResidencyLetterStatus,
    val purpose: String,
    @Json(name = "resident_name") val residentName: String,
    val address: ResidencyLetterAddress,
    /** Printed on the letter; what third parties verify. */
    @Json(name = "letter_code") val letterCode: String,
    @Json(name = "verify_url") val verifyUrl: String,
    @Json(name = "issued_at") val issuedAt: String,
    @Json(name = "revoked_at") val revokedAt: String? = null,
    @Json(name = "pdf_sha256") val pdfSha256: String,
)

/** `POST /api/homes/:id/residency-letters` body. */
@JsonClass(generateAdapter = true)
data class IssueResidencyLetterRequest(
    val purpose: String? = null,
)

/** `{ letter: … }` envelope (issue / revoke responses). */
@JsonClass(generateAdapter = true)
data class ResidencyLetterResponse(
    val letter: ResidencyLetter,
)

/** `{ letters: […] }` envelope (list response). */
@JsonClass(generateAdapter = true)
data class ResidencyLettersResponse(
    val letters: List<ResidencyLetter> = emptyList(),
)

/**
 * Public third-party check result — exactly what the paper shows.
 * Unknown codes come back as a uniform `{ valid: false }`.
 */
@JsonClass(generateAdapter = true)
data class ResidencyLetterVerification(
    val valid: Boolean,
    val status: ResidencyLetterStatus? = null,
    @Json(name = "resident_name") val residentName: String? = null,
    val address: ResidencyLetterAddress? = null,
    val purpose: String? = null,
    @Json(name = "issued_at") val issuedAt: String? = null,
    @Json(name = "revoked_at") val revokedAt: String? = null,
)

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Place/ResidencyLetterDTOs.swift
//
//  ResidencyLetterDTOs.swift
//  Pantopus
//
//  DTOs for `/api/homes/:id/residency-letters*` and the public
//  third-party check `/api/public/residency-letters/:code`.
//  Route: `backend/routes/residencyLetters.js`. Mirrors
//  `frontend/packages/api/src/endpoints/residencyLetters.ts`.
//
//  A T4 (verified-occupancy) resident issues a letter; the backend
//  freezes the printed facts + the exact PDF, and prints an unguessable
//  verification code on the letter. Letters are PERSONAL documents —
//  the API only ever returns the caller's own letters for a home.
//

import Foundation

public enum ResidencyLetterStatus: String, Sendable, Hashable {
    case issued
    case revoked
    case unknown
}

extension ResidencyLetterStatus: Decodable {
    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = ResidencyLetterStatus(rawValue: raw) ?? .unknown
    }
}

public struct ResidencyLetterAddress: Decodable, Sendable, Hashable {
    public let line1: String
    public let city: String?
    public let state: String?
    public let zipcode: String?
}

public struct ResidencyLetter: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let homeId: String
    public let status: ResidencyLetterStatus
    public let purpose: String
    public let residentName: String
    public let address: ResidencyLetterAddress
    /// Printed on the letter; what third parties verify.
    public let letterCode: String
    public let verifyUrl: String
    public let issuedAt: String
    public let revokedAt: String?
    public let pdfSha256: String

    private enum CodingKeys: String, CodingKey {
        case id, status, purpose, address
        case homeId = "home_id"
        case residentName = "resident_name"
        case letterCode = "letter_code"
        case verifyUrl = "verify_url"
        case issuedAt = "issued_at"
        case revokedAt = "revoked_at"
        case pdfSha256 = "pdf_sha256"
    }
}

/// `POST /api/homes/:id/residency-letters` body.
public struct IssueResidencyLetterRequest: Encodable, Sendable, Hashable {
    public let purpose: String?

    public init(purpose: String?) {
        self.purpose = purpose
    }
}

/// `{ letter: … }` envelope (issue / revoke responses).
public struct ResidencyLetterResponse: Decodable, Sendable, Hashable {
    public let letter: ResidencyLetter
}

/// `{ letters: […] }` envelope (list response).
public struct ResidencyLettersResponse: Decodable, Sendable, Hashable {
    public let letters: [ResidencyLetter]
}

/// Public third-party check result — exactly what the paper shows.
/// Unknown codes come back as a uniform `{ valid: false }`.
public struct ResidencyLetterVerification: Decodable, Sendable, Hashable {
    public let valid: Bool
    public let status: ResidencyLetterStatus?
    public let residentName: String?
    public let address: ResidencyLetterAddress?
    public let purpose: String?
    public let issuedAt: String?
    public let revokedAt: String?

    private enum CodingKeys: String, CodingKey {
        case valid, status, purpose, address
        case residentName = "resident_name"
        case issuedAt = "issued_at"
        case revokedAt = "revoked_at"
    }
}
